service/topology_coordinator.cc

```diff
@@ -740,6 +755,56 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
         }
     }
 
+    // If a node crashes after initiating gossip and before joining group0, it becomes orphan and 
+    // remains in gossip. This backgound fiber periodically checks for such nodes and purges those 
+    // entries from gossiper by commiting the node as left in raft.
+    future<> gossiper_orphan_remover_fiber() {
+        rtlogger.debug("start gossiper orphan remover fiber");
+
+        while (!_as.abort_requested()) {
+            try {
+                auto guard = co_await start_operation();
+                std::vector<canonical_mutation> updates;
+
+                _gossiper.for_each_endpoint_state([&](const gms::inet_address& addr, const gms::endpoint_state& eps) -> void {
+                    auto current_timestamp =
+                            std::chrono::duration_cast<std::chrono::seconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
+                    auto generation = eps.get_heart_beat_state().get_generation().value();
+                    static constexpr int32_t timeout = 60;
+                    auto host_id = eps.get_host_id();
+                    if (current_timestamp - generation > timeout && !_topo_sm._topology.contains(raft::server_id{host_id.id})) {
```

Comment:
> Assuming it never happens seems reasonable,

We are NOT assuming it never happens. In this case, since the topology operations are sequential due to guards, node removal and join_group0 wont cause race condition, and if 1st one wins, the node is banned and you need to restart with new host id, other wise, in second case, the banning would be halted because the node is already joined.

>We could use 2 * ring_delay + 5 s instead of 60 s

Lets not couple this simple logic with something else, the number you proposed seems brittle, and changing the ring delay can silently lead to changing our timeout.